go/cmd/dosage-filter/main.go

```diff
@@ -0,0 +1,316 @@
+
+	flag.StringVar(&cliargs.outputPath, "output", "", "The path to the output Arrow IPC/Feather file")
+	flag.StringVar(&cliargs.outputPath, "out", "", "The path to the output Arrow IPC/Feather file")
+
+	flag.StringVar(&cliargs.beanstalkConfigPath, "queue-config", "", "The path to the Beanstalkd queue connection config (e.g. config/beanstalk.yml)")
+	flag.StringVar(&cliargs.beanstalkConfigPath, "q", "", "The path to the Beanstalkd queue connection config (short form)")
+	flag.Int64Var(&cliargs.progressFrequency, "progress-frequency", int64(5e3), "Print progress every N variants processed")
+	flag.Int64Var(&cliargs.progressFrequency, "p", int64(5e3), "Print progress every N variants processed (short form)")
+
+	flag.StringVar(&cliargs.jobSubmissionID, "job-submission-id", "", "The job submission ID")
+	flag.StringVar(&cliargs.jobSubmissionID, "j", "", "The job submission ID (short form)")
+
+	a := os.Args[1:]
+	if args != nil {
+		a = args
+	}
+
+	flag.CommandLine.Parse(a)
+
+	if err := validateArgs(cliargs); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	return cliargs
+}
+
+// validateArgs checks if the required arguments are provided.
+func validateArgs(args *CLIArgs) error {
+	var missing []string
+	v := *args
+	if v.inputPath == "" {
+		missing = append(missing, "input")
+	}
+	if v.outputPath == "" {
+		missing = append(missing, "output")
+	}
+	if v.lociPath == "" {
+		missing = append(missing, "loci")
+	}
+
+	if v.jobSubmissionID == "" {
+		if v.beanstalkConfigPath != "" {
+			missing = append(missing, "job-submission-id")
+		}
+	}
+
+	if len(missing) > 0 {
+		return fmt.Errorf("missing required arguments: %s", strings.Join(missing, ", "))
+	}
+
+	return nil
+}
+
+// readLociFile reads a file containing loci to filter and returns a map of loci.
+// Each line in the file represents a locus.
+func readLociFile(filePath string) (map[string]bool, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	loci := make(map[string]bool)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		loci[scanner.Text()] = true
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return loci, nil
+}
+
+// processRecordAt processes a record at the given index, filters the rows based on the loci,
+// and writes the filtered rows to the arrowWriter.
+func processRecordAt(fr *ipc.FileReader, loci map[string]bool, arrowWriter *bystroArrow.ArrowWriter, queue chan int, complete chan bool, count *atomic.Int64) {
+	pool := memory.NewGoAllocator()
+	builder := array.NewRecordBuilder(pool, arrowWriter.Schema)
+	defer builder.Release()
+
+	rowsAccumulated := 0
+
+	for index := range queue {
+		record, err := fr.RecordAt(index)
+		if err != nil {
+			log.Fatalf("Failed to read record at %d: %v", index, err)
+		}
+
+		locusCol := record.Column(0).(*array.String)
```

Comment:
the locus column is always the first column, see bystro-vcf; in future iterations we could generalize this, but I don't see the need since we control the format.